README
NAME
    pler - The DWIM Perl Debugger

DESCRIPTION
    pler is a small script which provides a sanity layer for debugging test
    scripts in Perl distributions.

    While prove has proven itself to be a highly useful program for manually
    running one or more groups of scripts in a distribution, what we also
    need is something that provides a similar level of intelligence in a
    debugging context.

    pler checks that the environment is sound, runs some cleanup tasks if
    needed, makes sure you are in the right directory, and then hands off to
    the perl debugger as normal.

    Command-line arguments are assumed to be test files or substrings of
    filenames. pler will search through the t/ and xt/ subdirectories and
    test all matching files.

    Arguments can be passed to the test file(s) on the command line after a
    --.

TO DO
    - Tweak some small terminal related issues on Win32

SUPPORT
    All bugs should be filed via the bug tracker at

    <http://rt.cpan.org/NoAuth/ReportBug.html?Queue=pler>

    For other issues, or commercial enhancement and support, contact the
    author

AUTHOR
    Adam Kennedy <adamk@cpan.org>

SEE ALSO
    prove, <http://ali.as/>

COPYRIGHT
    Copyright 2006 - 2013 Adam Kennedy.

    This program is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.

    The full text of the license can be found in the LICENSE file included
    with this module.



Changes
Revision history for Perl extension pler

1.07 Mon 30 Sep 2013
	- Allow command-line arguments to be passed to test files (jkeroes)

1.06 Thu 10 Jun 2010
	- Apply rel2abs to the PERL5LIB additions so that path-changing
	  tests can still load modules properly.

1.05 Sun 24 Jan 2010
	- When executing the test script, enable RELEASE_TESTING

1.04 Wed 20 Jan 2010
	- Use a -I to PERL5LIB conversion trick similar to the it is done
	  by TAP::Parser::Source::Perl.
	- Add the inc directory to the -I list, as MakeMaker does.

1.03 Tue 19 Jan 2010
	- Removing the destructive Carp 0 dependency

1.02 Tue  5 Jan 2010
	- Make the integer shortcut more flexible, so it can handle
	  more than just an underscore after the numbers.

1.01 Sun  3 Jan 2010
	- Fixed a silly typo

1.00 Sun  3 Jan 2010
	- Switching to a production version
	- Upgraded to Module::Install::DSL 0.91
	- Added support for the xt directory
	- Added support for specialised integer filter
	- Added some actual tests for the filter logic

0.34 Mon 30 Jun 2008
	- Optimising file search to half the number of stat
	  calls that need to be made to the system.

0.33 Fri 18 Apr 2008
	- Enable HARNESS_ACTIVE when calling scripts

0.32 Tue  1 Jan 2008
	- Replacing the use of $^X with Probe::Perl

0.31 Thu 27 Sep 2007
	- Fixing multiple filters
	- Incrementing to get new author tests
	- Upgrading to Module::Install 0.68

0.30 Thu  6 Sep 2007
	- Adding support for multiple filters
	- Detect the distroot correctly from multiple
	  levels down subdirectories.

0.29 Fri 14 Sep 2007
	- Bug in multiple filters would cause an infinite loop, fixed
	- Fix a mistake in the version mismatch error message

0.28 Fri 14 Sep 2007
	- Added support for multiple filters, to make life easier in big test suites
	  (So stuff like "pler db 4" will match t/db-tests/04_foo.t but not t/04_static.t)
	- All pattern filters are now case-insensitive
	- Bug Fix: Stopped needlessly running ./Build every time
	  (A typo meant I was statting the wrong file in the shortcutting test)
	- Script now tests that it loads the module with matching version
	- Adding explicit dependencies for core modules (just in case)

0.27 Mon 27 Aug 2007
	- Added a first experimental attempt at Module::Build support

0.26 Mon 27 Aug 2007
	- Adding a -V option for the version
	- Adding a --help option for help/usage
	- Adding -w which will be passed through to the underlying perl call

0.25 Tue 21 Aug 2007
	- When inside the t directory and the user uses tab-completion to specify the
	  filename, call the file they meant even after we chdir up to the root.
	- If the root has a Makefile, but NOT a Makefile.PL, don't run it and risk
	  causing some unknown functionality to run.

0.24 Tue 21 Aug 2007
	- I am an idiot, and didn't actually USE the Cwd::chdir in the below release
	  (this fixes that)

0.23 Tue 21 Aug 2007
	- Use Cwd::chdir instead of the core chdir so that $ENV{PWD} is kept up to date
	  (This prevents pler breaking code which uses $ENV{PWD})
	- Adding a more recent dependency on 3.00 Pathtools to make sure we have Cwd

0.22 Tue 21 Aug 2007
	- If no Makefile.PL, intuit the dist root by looking for a "lib" and "t" dir

0.21 Mon 20 Aug 2007
	- dumb is spelled with a "b", not a "p". Oh the irony :(

0.20 Thu  2 Aug 2007
	- Moving bin/pler to the more standard script/pler
	- Collapse Devel::Pler into pler.pm
	- Adding support for regenerating the Makefile from Makefile.PL
	  (only on Win32, because the Makefile is not capable of regenerating)
	- Set TERM=dumb if no TERM value is set
	  (prevents a noisy error when the debugger starts up)

0.19 skipped

0.18 Wed 16 May 2007
        - Adding the ability to run scripts from the t directory
        - Moving pler.pm to Devel::Pler
        - Moving bin/pler to pler.pm
        - Added a stubbed launch script as bin/pler

0.17 Fri 11 May 2007
        - Fixing a path bug that breaks Test::Files
	- Updated Test::Script dependency to 1.02

0.16 Tue  2 Jan 2007
	- Fixed a test bug I thought I'd fixed already

0.15 Mon  1 Jan 2007
	- Converted from Devel::Pler to just pler.pm to allow
	  for better installation and documentation on search.cpan.
	- You can now "cpan -i pler"

0.14 Fri 15 Sep 2006
	- exec() does not work on Win32 (and we assume cygwin).
	  Instead, use system() which uses more memory, but
	  does at least work.

0.13 Thu  3 Aug 2006
	- Bug fixes
	  (I honestly don't know his this crap gets out to CPAN)

0.12 Thu 25 Jul 2006
	- Show the list of scripts with no params, or run script
	  if there is only one in the t directory.
	- When verbose is off, comment on the script name for clarity
	- Copying the POD in the Devel::Pler for better toolchain
	  compatibility.

0.11 missing

0.10 Fri 21 Jul 2006
	- Added support for refering to scripts by a unique substring

0.09 Thu 20 Jul 2006
	- Left an ADAMK::Debug reference in 01_compile.t

0.08 Wed 19 Jul 2006
	- Bug fix to remove remaining ADAMK::Debug references
	- Fix some other grammar/spelling errors.

0.07 Wed 19 Jul 2006
	- Renaming apld script to pler :)
	- Renaming ADAMK::Debug to Devel::Pler
	- Renaming distribution from ADAMK-Debug to Devel-Pler

0.06 Wed 19 Jul 2006
	- Add lib to @INC if there is not blib but there is an inc

0.05 Tue 18 Jul 2006
	- Bug fix

0.04 Tue 18 Jul 2006
	- This time, make ADAMK::Debug an actual Exporter

0.03 Tue 18 Jul 2006
	- Call perl -d with -Mblib if needed
	- Enable warnings in apld by default

0.02 Tue 18 Jul 2006
	- Run make if needed
	- Shifted support functions out to the module

0.01 Tue 18 Jul 2006
	- Created the initial implementation


lib/pler.pm
package pler;

# See 'sub main' for main functionality

use 5.00503;
use strict;
use Config;
use Carp                       ();
use Cwd                   3.00 ();
use File::Which           0.05 ();
use File::Spec            0.80 ();
use File::Spec::Functions      ':ALL';
use File::Find::Rule      0.20 ();
use Getopt::Long             0 ();
use Probe::Perl           0.01 ();

use vars qw{$VERSION};
BEGIN {
		$VERSION = '1.07';
}

# Does exec work on this platform
use constant EXEC_OK => ($^O ne 'MSWin32' and $^O ne 'cygwin');

# Can you overwrite an open file on this platform
use constant OVERWRITE_OK => !! ( $^O ne 'MSWin32' );






#####################################################################
# Resource Locations

sub MakefilePL () {
	catfile( curdir(), 'Makefile.PL' );
}

sub BuildPL () {
	catfile( curdir(), 'Build.PL' );
}

sub Makefile () {
	catfile( curdir(), 'Makefile' );
}

sub Build () {
	catfile( curdir(), 'Build' );
}

sub perl () {
	Probe::Perl->find_perl_interpreter;
}

# Look for make in $Config
sub make () {
	my $make  = $Config::Config{make};
	my $found = File::Which::which( $make );
	unless ( $found ) {
		Carp::croak("Failed to find '$make' (as specified by \$Config{make})");
	}
	return $found;
}

sub blib () {
	catdir( curdir(), 'blib' );
}

sub inc () {
	catdir( curdir(), 'inc' );
}

sub lib () {
	catdir( curdir(), 'lib' );
}

sub t () {
	catdir( curdir(), 't' );
}

sub xt () {
	catdir( curdir(), 'xt' );
}





#####################################################################
# Convenience Logic

sub has_makefilepl () {
	!! -f MakefilePL;
}

sub has_buildpl () {
	!! -f BuildPL;
}

sub has_makefile () {
	!! -f Makefile;
}

sub has_build () {
	!! -f Build;
}

sub has_blib () {
	!! -d blib;
}

sub blibpm () {
	eval {
		require blib;
	};
	return ! $@;
}

sub has_inc () {
	!! -f inc;
}

sub has_lib () {
	!! -d lib;
}

sub has_t () {
	!! -d t;
}

sub has_xt () {
	!! -d xt;
}

sub in_distroot () {
	!! (
		has_makefilepl or (has_lib and has_t)
	);
}

sub in_subdir () {
	!! (
		-f catfile( updir(), 'Makefile.PL' )
		or
		-d catdir( updir(), 't' )
	);
}

sub needs_makefile () {
	has_makefilepl and ! has_makefile;
}

sub needs_build () {
	has_buildpl and ! has_build;
}

sub mtime ($) {
	(stat($_[0]))[9];
}

sub old_makefile () {
	has_makefile
	and
	has_makefilepl
	and
	mtime(Makefile) < mtime(MakefilePL);
}

sub old_build () {
	has_build
	and
	has_buildpl
	and
	mtime(Build) < mtime(BuildPL);
}





#####################################################################
# Utility Functions

# Support verbosity
use vars qw{$VERBOSE};
BEGIN {
	$VERBOSE ||= 0;
}

sub is_verbose {
	$VERBOSE;
}

sub verbose ($) {
	message( $_[0] ) if $VERBOSE;
}

sub message ($) {
		print $_[0];
}

sub error (@) {
	print ' ' . join '', map { "$_\n" } ('', @_, '');
	exit(255);
}

sub run ($) {
	my $cmd = shift;
	verbose( "> $cmd" );
	system( $cmd );
}

sub handoff (@) {
	my $cmd = join ' ', @_;
	verbose( "> $cmd" );
	$ENV{HARNESS_ACTIVE}  = 1;
	$ENV{RELEASE_TESTING} = 1;
	if ( EXEC_OK ) {
		exec( @_ ) or Carp::croak("Failed to exec '$cmd'");
	} else {
		system( @_ );
		exit(0);
	}
}


#####################################################################
# Main Script

my @SWITCHES = ();

sub main {
	Getopt::Long::Configure('no_ignore_case');
	Getopt::Long::GetOptions(
		'help' => \&help,
		'V'    => sub { print "pler $VERSION\n"; exit(0) },
		'w'    => sub { push @SWITCHES, '-w' },
	);

	# Get the script name
	my $script = shift @ARGV;
	unless ( defined $script ) {
		print "# No file name pattern provided, using 't'...\n";
		$script = 't';
	}

	# Abuse the highly mature logic in Cwd to define an $ENV{PWD} value
	# by chdir'ing to the current directory.
	# This lets us get the current directory without losing symlinks.
	Cwd::chdir(curdir());
	my $orig = $ENV{PWD} or die "Failed to get original directory";

	# Can we locate the distribution root
	my ($v,$d,$f) = splitpath($ENV{PWD}, 'nofile');
	my @dirs      = splitdir($d);
	while ( @dirs ) {
		my $buildpl = catpath(
			$v, catdir(@dirs), BuildPL,
		);
		my $makefilepl = catpath(
			$v, catdir(@dirs), MakefilePL,
		);
		unless ( -f $buildpl or -f $makefilepl ) {
			pop @dirs;
			next;
		}

		# This is a distroot
		my $distroot = catpath( $v, catdir(@dirs), "" );
		Cwd::chdir($distroot);
		last;
	}
		unless ( in_distroot ) {
				error "Failed to locate the distribution root";
		}

	# Makefile.PL? Or Build.PL?
	my $BUILD_SYSTEM = has_buildpl ? 'build' : has_makefilepl ? 'make' : '';
	if ( $BUILD_SYSTEM eq 'build' ) {
		# Because Module::Build always runs with warnings on,
		# pler will as well when you use a Build.PL
		unless ( grep { $_ eq '-w' } @SWITCHES ) {
			push @SWITCHES, '-w';
		}
	}

	# If needed, regenerate the Makefile or Build file
	# Currently we do not remember Makefile.PL or Build.PL params
	if ( $BUILD_SYSTEM eq 'make' ) {
		if ( needs_makefile or (old_makefile and ! OVERWRITE_OK) ) {
			run( join ' ', perl, MakefilePL );
		}
	} elsif ( $BUILD_SYSTEM eq 'build' ) {
		if ( needs_build or old_build ) {
			run( join ' ', perl, BuildPL );
		}
	}

	# Locate the test script to run
	if ( $script =~ /\.t$/ ) {
		# EITHER
		# 1. They tab-completed the script relative to the original directory (most likely)
		# OR
		# 2. They typed the entire name of the test script
		my $tab_completed = File::Spec->catfile( $orig, $script );
		if ( -f $tab_completed ) {
			if ( $orig eq $ENV{PWD} ) {
				$script = $script; # Included for clarity
			} else {
				$script = File::Spec->abs2rel( $tab_completed, $ENV{PWD} );
			}
		}

		} else {
		# Get the list of possible tests
		my @directory = ( 't', has_xt ? 'xt' : () );
		my @possible  = File::Find::Rule->name('*.t')->file->in(@directory);

		# Filter by the search terms to find matching tests
		my $matches = filter(
			[ $script, @ARGV ],
			[ @possible ],
		);
		unless ( @$matches ) {
			error "No tests match '$script'";
		}
		if ( @$matches > 1 ) {
			error(
					"More than one possible test",
					map { "  $_" } sort @$matches,
			);
		}
		$script = $matches->[0];

		# Localize the path
		$script = File::Spec->catfile( split /\//, $script );
	}
	unless ( -f $script ) {
		error "Test script '$script' does not exist";
	}

		# Rerun make or Build if needed
	if ( $BUILD_SYSTEM eq 'make' ) {
		# Do NOT run make if there is no Makefile.PL, because it likely means
		# there is a hand-written Makefile and NOT one derived from Makefile.PL,
		# and we have no idea what functionality we might trigger.
			if ( in_distroot and has_makefile and has_makefilepl ) {
					run( make );
			}
	} elsif ( $BUILD_SYSTEM eq 'build' ) {
		if ( in_distroot and has_build and has_buildpl ) {
			run( Build );
		}
	}

	# Passing includes via -I params is not good enough
	# because you can't subshell them, and it's also not
	# how MakeMaker does it anyway.
	# We need to hack/extend PERL5LIB instead.
	my $path_sep = $Config{path_sep};
	my @PERL5LIB = ();

	# Build the command to execute
	my @flags = @SWITCHES;
	if ( has_blib ) {
		if ( has_inc ) {
			push @PERL5LIB, inc;
		}
		push @PERL5LIB, File::Spec->catdir(
			blib, 'lib',
		);
		push @PERL5LIB, File::Spec->catdir(
			blib, 'arch',
		);
	} elsif ( has_lib ) {
		push @PERL5LIB, lib;
	}

	# Absolutify the PERL5LIB elements so they will survive
	# the test script changing it's CWD. This was added to
	# deal with the path-shifting of the Padre tests.
	@PERL5LIB = map {
		File::Spec->rel2abs($_)
	} @PERL5LIB;

	# Hand off to the perl debugger
	unless ( pler->is_verbose ) {
		message( "# Debugging $script...\n" );
	}
	my @cmd = ( perl, @flags, '-d', $script, @ARGV );
	local $ENV{PERL5LIB} = defined($ENV{PERL5LIB})
		? join( $path_sep, @PERL5LIB, $ENV{PERL5LIB} )
		: join( $path_sep, @PERL5LIB );
	handoff( @cmd );
}

# Encapsulates the smart filtering as a function
sub filter {
	my $terms    = shift;
	my $possible = shift;
	my @matches  = @$possible;

	while ( @$terms ) {
		my $term = shift @$terms;

		if ( ref $term eq 'Regexp' ) {
			# If the term is a regexp apply it directly
			@matches = grep { $_ =~ $term } @matches;
		} elsif ( $term =~ /^[1-9]\d*$/ ) {
			# If the search is a pure integer (without leading
			# zeros) attempt a specialised numeric filter.
			@matches = grep { /\b0*${term}[^0-9]/ } @matches;
		} else {
			# Otherwise treat it as a naive string match
			$term = quotemeta $term;
			@matches = grep { /$term/i } @matches;
		}
	}

	return \@matches;
}

sub help { print <<'END_HELP'; exit(0); }
Usage:
	pler [options] [file/pattern] [-- arguments for test files]

Options:
		-V              Print the pler version
		-h, --help      Display this help
		-w              Run test with the -w warnings flag
END_HELP

1;

=pod

=head1 NAME

pler - The DWIM Perl Debugger

=head1 DESCRIPTION

B<pler> is a small script which provides a sanity layer for debugging
test scripts in Perl distributions.

While L<prove> has proven itself to be a highly useful program for
manually running one or more groups of scripts in a distribution,
what we also need is something that provides a similar level of
intelligence in a debugging context.

B<pler> checks that the environment is sound, runs some cleanup tasks
if needed, makes sure you are in the right directory, and then hands off
to the perl debugger as normal.

Command-line arguments are assumed to be test files or substrings of
filenames. pler will search through the t/ and xt/ subdirectories and
test all matching files.

Arguments can be passed to the test file(s) on the command line after
a --.

=head1 TO DO

- Tweak some small terminal related issues on Win32

=head1 SUPPORT

All bugs should be filed via the bug tracker at

L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=pler>

For other issues, or commercial enhancement and support, contact the author

=head1 AUTHOR

Adam Kennedy E<lt>adamk@cpan.orgE<gt>

=head1 SEE ALSO

L<prove>, L<http://ali.as/>

=head1 COPYRIGHT

Copyright 2006 - 2013 Adam Kennedy.

This program is free software; you can redistribute
it and/or modify it under the same terms as Perl itself.

The full text of the license can be found in the
LICENSE file included with this module.

=cut
